### ISSUE TITLE
intelliSenseOptions and additionalCompilerArgs documentation is completely unclear

### ISSUE THREAD

**Issue by robindegen** [MicrosoftDocs/cpp-docs#3750]
The documentation on the cmake presets json mentions the possible use of intelliSenseOptions and additionalCompilerArgs to feed intellisense information (supposedly about include paths and defines?) But it is completely unclear how this works. It is also unclear if it is assumed that linux/mac/windows and clang is expected for this to work at all.

I'm currently attempting to use emscripten (clang-based) from Visual Studio and I have no idea how to configure these fields to give intellisense the information that it needs (mostly just include search paths).

There are no examples, no context. It plainly states which "additional arguments" can be given, and if it should be the defaults? What does that mean? Can I also just provide paths to intellisense myself? There is no mention of this.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 64d317ae-6e03-8bb6-2b2c-4d1b9ecf3080
* Version Independent ID: cab1289c-b480-e674-106f-f37d3a3f198f
* Content: [CMakeUserPresets.json](https://docs.microsoft.com/en-us/cpp/build/cmake-presets-json-reference?view=msvc-170)
* Content Source: [docs/build/cmake-presets-json-reference.md](https://github.com/Microsoft/cpp-docs/blob/main/docs/build/cmake-presets-json-reference.md)
* Product: **visual-cpp**
* Technology: **cpp-tools**
* GitHub Login: @corob-msft
* Microsoft Alias: **corob**

**Comment by corob-msft** [comment]
@robindegen 

Sorry you're not finding the information you're looking for. As far as I can tell, the `additionalCompilerArgs` value allows arbitrary compiler arguments. There's no schema for those, and the values depend on the platform and compiler in use. If you're using cl.exe, presumably you could specify [`/I`](https://docs.microsoft.com/en-us/cpp/build/reference/i-additional-include-directories?view=msvc-170) or [`/D`](https://docs.microsoft.com/en-us/cpp/build/reference/d-preprocessor-definitions?view=msvc-170) arguments to specify specific include paths and preprocessor macros.  @sinemakinci1 may be able to elaborate further, or correct me if I'm wrong.

**Comment by robindegen** [comment]
@corob-msft Thank you for your response. I think this already kinda indicates how unclear it is. Because according to this blog post it's something completely different:
https://devblogs.microsoft.com/cppblog/configure-intellisense-with-cmake-toolchain-files-in-visual-studio-2019-16-9-preview-2/

```
additionalCompilerArgs can be used to explicitly add arguments to be used when Visual Studio queries the compiler for IntelliSense. useCompilerDefaults can be used to disable the use of the GCC/Clang compiler for IntelliSense configuration (defaults to true). These options apply to Windows, Linux, and WSL configurations.
```

It mentions that it's for feeding information to Intellisense. Not the compiler itself. This would be done by Cmake I would expect. I can imagine there being a system where Intellisense would query the compiler used to figure out what all the system paths are (sysroot), so that it knows where the stl headers are etc. But if that is true, then there is no information about how it normally queries this information, nor what kind of output it expects from the compiler when this is called. I can imagine most clang/gcc versions will give a rather standardized output; but this is not always the case.

**Comment by corob-msft** [comment]
@robindegen 

The IntelliSense system in Visual Studio is driven by a separate compiler invoked by the IDE. When IntelliSense runs, there's no "querying of the compiler" taking place. The IntelliSense engine is told the include path and preprocessor definition values directly by the IDE. The IDE normally passes it the same values for the include search paths and preprocessor definitions that it passes to the regular compiler toolchain. The form these parameters take is an internal detail that's not usually exposed. My assumption was that these values would take the same form as whatever you'd pass to your regular toolchain. I might have been wrong about that, given the statement in the `useCompilerDefaults` section about using gcc-style arguments. So, you'd need to use the `-I` and `-D` forms, instead of `/I` and `/D`, as a common form which will work for cl, clang, and gcc. Something like `-isystem` won't work with cl, but presumably will work with gcc and clang.

Just to clarify when to use this property: If you want to send the IntelliSense engine *different values* than your normal toolchain uses, then `additionalCompilerArgs` is the way to do it. If you want it to use the default values set for your project, you don't need to set anything in `additionalCompilerArgs` at all.

There are other ways the IntelliSense display can go wrong. If your builds succeed but IntelliSense isn't working, it's probably not because you're setting `additionalCompilerArgs` incorrectly. For example, if your toolchain references environment variables such as `CPLUS_INCLUDE_PATH` that aren't set as part of the CMake project configuration, you might find they also aren't picked up by the IntelliSense engine.

**Comment by sykhro** [comment]
> The documentation on the cmake presets json mentions the possible use of intelliSenseOptions and additionalCompilerArgs to feed intellisense information (supposedly about include paths and defines?) But it is completely unclear how this works. It is also unclear if it is assumed that linux/mac/windows and clang is expected for this to work at all.
> 
> I'm currently attempting to use emscripten (clang-based) from Visual Studio and I have no idea how to configure these fields to give intellisense the information that it needs (mostly just include search paths).
> 
> There are no examples, no context. It plainly states which "additional arguments" can be given, and if it should be the defaults? What does that mean? Arguments to what? The compiler, or Intellisense itself? What if the compiler provides no such functionality? Can I also just provide paths to intellisense myself? There is no mention of this.
> 
> #### Document Details
> ⚠ _Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking._
> 
> * ID: 64d317ae-6e03-8bb6-2b2c-4d1b9ecf3080
> * Version Independent ID: cab1289c-b480-e674-106f-f37d3a3f198f
> * Content: [CMakeUserPresets.json](https://docs.microsoft.com/en-us/cpp/build/cmake-presets-json-reference?view=msvc-170)
> * Content Source: [docs/build/cmake-presets-json-reference.md](https://github.com/Microsoft/cpp-docs/blob/main/docs/build/cmake-presets-json-reference.md)
> * Product: **visual-cpp**
> * Technology: **cpp-tools**
> * GitHub Login: @corob-msft
> * Microsoft Alias: **corob**

Were you ever able to get this working? I'm facing the same issues. I've dumped the cflags using `em++ --cflags` and have configured the settings as follows, doesn't seem to actually do anything:
```json
"microsoft.com/VisualStudioSettings/CMake/1.0": {
  "intelliSenseMode": "windows-clang-x64",
  "intelliSenseOptions": {
    "useCompilerDefaults": false,
    "additionalCompilerArgs": [
      "-target",
      "wasm32-unknown-emscripten",
      "-fignore-exceptions",
      "-fvisibility=default",
      "-mllvm",
      "-combiner-global-alias-analysis=false",
      "-mllvm",
      "-enable-emscripten-sjlj",
      "-mllvm",
      "-disable-lsr",
      "--sysroot=<path/to/upstream/emscripten/cache/sysroot>",
      "-DEMSCRIPTEN",
      "-Xclang",
      "-iwithsysroot/include/fakesdl",
      "-Xclang",
      "-iwithsysroot/include/compat"
    ]
  }
}
```

**Comment by robindegen** [comment]
I never did. The fact that Microsoft decided to have a product called "Visual Studio Code" with a json file that has the exact same name as the regular Visual Studio one, but in a slightly different format; REALLY isn't helping all this. Googling for issues in regular Visual Studio has become a nightmare. Everything is poluted with tons of people proposing vscode solutions.

**Comment by sykhro** [comment]
> I never did. The fact that Microsoft decided to have a product called "Visual Studio Code" with a json file that has the exact same name as the regular Visual Studio one, but in a slightly different format; REALLY isn't helping all this. Googling for issues in regular Visual Studio has become a nightmare. Everything is polluted with tons of people proposing vscode solutions.

I agree with the sentiment. The issue seems to be the `--sysroot=` flag, using `-isystem` or even `-isysroot` it works. Same  when using clangd.
